File: utils/aoc_utils.py
from typing import Any, Callable, Union, Optional, Tuple
import pathlib
import datetime
import re
from enum import Enum
from os.path import exists
from datetime import timedelta

TestFunction = Union[
    Callable[[list[Any]], Optional[str | int]],
    Callable[[str], Optional[str | int]],
    Callable[[int], Optional[str | int]],
]


AnswerFunction = Union[
    Callable[[list[Any]], Optional[str | int]],
    Callable[[str], Optional[str | int]],
    Callable[[int], Optional[str | int]],
]
FormFunction = Callable[[list[str]], Any]
Map = list[list[str]]


class RegexException(Exception):
    pass


class ParsingException(Exception):
    pass


class Parser:
    def parse(self, line: str, strip: bool = True) -> Any:
        if strip:
            return line.strip()

        return line


class OneLineParser(Parser):
    def parse(self, line: str, strip: bool = True) -> Any:
        if strip:
            return line.strip()

        return line


class IntParser(Parser):
    def parse(self, line: str, strip: bool = True) -> Any:
        return int(line.strip())


class MapParser(Parser):
    def __init__(self, integer=False):
        self.integer = integer

    def parse(self, line, strip: bool = True) -> list[int | str]:
        if self.integer:
            return [int(ch) for ch in line.strip()]
        else:
            return [ch for ch in line.strip()]


class RegexParser(Parser):
    def __init__(self, regexes: list[tuple[str, Optional[FormFunction]]]):
        self.regexes = [(re.compile(regex[0]), regex[1]) for regex in regexes]

    def parse(self, line, strip: bool = True) -> Any:
        for regex, form in self.regexes:
            if m := regex.match(line):
                if form:
                    return form(list(m.groups()))
                else:
                    return m.groups()

        raise RegexException(f"Could not parse '{line}'")

    def _form(self, results: tuple) -> Any:
        return results


class Day:
    def __init__(
        self,
        year: int,
        day: int,
        test_one: Optional[TestFunction],
        test_two: Optional[TestFunction],
        part_one: AnswerFunction,
        part_two: AnswerFunction,
        test_input: Optional[list[str] | str | int | Parser] = None,
        input: list[str] | str | int | Parser = Parser(),
    ):
        self.year = year
        self.day = day
        self.dir = pathlib.Path(f"Advent{self.year}", f"Day{self.day}")
        self.input = input
        self.test_input = test_input if test_input else self.input
        self.test_one = test_one
        self.test_two = test_two
        self.part_one = part_one
        self.part_two = part_two

    def multi_line_input(self, filename="input.txt", parser: Parser = Parser()):
        try:
            rc = []

            for i, line in enumerate(open(self.dir / filename)):
                if (parsed := parser.parse(line)) != None:
                    rc.append(parsed)
            return rc

        except RegexException as ex:
            raise ParsingException(
                f"Parse filed in {filename} on line {i}: {ex}"
            ) from ex

    def header(self):
        print()
        print(f"Advent of code for Year {self.year} Day {self.day}:")
        print()

    def runner(
        self, fn, infile: str, input_method: str | int | list[str] | Parser
    ) -> tuple[Any, timedelta]:
        match input_method:
            case [*input]:
                for line in input:
                    start = datetime.datetime.now()
                    answer = fn(line)
                    elapsed = datetime.datetime.now() - start
                    print(
                        f"({elapsed}) Test: {answer if answer is not None else 'None':<10} <- [ {line} ]"[
                            :100
                        ]
                    )
                answer = None

            case str(input) | int(input):
                start = datetime.datetime.now()
                answer = fn(input)
                elapsed = datetime.datetime.now() - start

            case OneLineParser():
                start = datetime.datetime.now()
                answer = fn(
                    self.multi_line_input(filename=infile, parser=input_method)[0]
                )
                elapsed = datetime.datetime.now() - start

            case Parser():
                start = datetime.datetime.now()
                answer = fn(self.multi_line_input(filename=infile, parser=input_method))
                elapsed = datetime.datetime.now() - start

            case _:
                raise ParsingException(f"Unknown data input type {input_method}")
        return answer, elapsed

    def test_it(self, fn, filename="test_input.txt") -> None:
        if fn:
            answer, elapsed = self.runner(fn, filename, self.test_input)
            if answer:
                print(f"({elapsed}) Test is {answer if answer else 'I Dunno'}")

    def run_it(self, fn, name: str) -> None:
        answer, elapsed = self.runner(fn, "input.txt", self.input)
        print(f"({elapsed}) Answer for {name} is {answer if answer else 'I Dunno'}")

    def run_all(self, run_tests=False):
        self.header()
        if run_tests:
            self.test_it(self.test_one)
        self.run_it(self.part_one, "Part One")
        print()
        if run_tests:
            self.test_it(
                self.test_two,
                filename="test_input2.txt"
                if exists(self.dir / "test_input2.txt")
                else "test_input.txt",
            )
        self.run_it(self.part_two, "Part Two")

File: utils/aoc_utils.cpp
#include <iostream>
#include <fstream>
#include <iterator>
#include <iomanip>
#include <string>
#include <vector>
#include <algorithm>
#include <numeric>
#include <chrono>

#include <cstdio>

#include "aoc_utils.h"

using namespace std;
using namespace std::chrono;

namespace au
{

    void *OldenDay::read_file(const std::string &filename, char **input, unsigned int &length)
    {
        FILE *fp = fopen(filename.c_str(), "rt");
        fseek(fp, 0, SEEK_END);
        unsigned int bytes = ftell(fp);

        fseek(fp, 0, SEEK_SET);

        char *buffer = new char[bytes];
        *input = buffer;
        const size_t rc = fread(buffer, 1, bytes, fp);

        length = 0;
        bool ending = true;
        for (unsigned int i = 0; i < bytes; i++)
        {
            if (buffer[i] == '\n' || buffer[i] == '\r')
            {
                buffer[i] = '\0';
                length++;
            }
        }
        length++;
        char **data = new char *[length];
        unsigned line = 0;
        for (unsigned int i = 0; i < bytes; i++)
            if (i == 0 || buffer[i - 1] == '\0')
                data[line++] = buffer + i;

        // Handle last line being empty
        if (data[length - 1] == 0)
            length -= 1;

        if (m_parser)
            return m_parser->parse((void *)data, length);

        return (void *)data;
    }

    void OldenDay::runner(test_funcptr fp, const char *text, void *input, unsigned int length)
    {
        auto start = high_resolution_clock::now();
        Answer answer = (this->*fp)(input, length);
        auto time_span = duration_cast<duration<double>>(high_resolution_clock::now() - start);
        printf("(%5.9f) %s %s\n", time_span.count(), text, answer.text());
    }

    void OldenDay::run_all(bool run_tests)
    {
        unsigned int test_length, input_length;
        void *test_input = read_file("test_input.txt", &testbuffer, test_length);
        void *input = read_file("input.txt", &inputbuffer, input_length);

        printf("Advent of code for Year %d Day %d\n\n", m_year, m_day);

        test_funcptr ptest_one = &OldenDay::test_one;
        if (run_tests)
            runner(&OldenDay::test_one, "Test is", test_input, test_length);

        runner(&OldenDay::part_one, "Answer for Part One is", input, input_length);

        printf("\n");

        if (!m_parser)
        {
            delete[] (char **)test_input;
            delete[] (char **)input;
        }
        test_input = read_file("test_input.txt", &testbuffer, test_length);
        input = read_file("input.txt", &inputbuffer, input_length);
        if (run_tests)
            runner(&OldenDay::test_two, "Test is", test_input, test_length);

        runner(&OldenDay::part_two, "Answer for Part Two is", input, input_length);

        if (!m_parser)
        {
            delete[] (char **)test_input;
            delete[] (char **)input;
        }
    }
}
